### PR TITLE
Security permissions for Groovy closures

### LINKED ISSUE #16194: Groovy Closures inaccessible
This starts with ES 2.2 as we move to the script modules. The below script works in ES 2.1, but not ES 2.2 (the impact of which could be pretty big):

```http
PUT /test/type/1
{
  "message": "This is a message!",
  "numbers" : [1, 2, 3, 4]
}

PUT /test/type/2
{
  "message": "This is a message! too",
  "numbers" : [4, 2, 3, 5]
}

GET /test/_search
{
  "script_fields": {
    "use_closure": {
      "script": {
        "inline": "doc['numbers'].values.findAll { it % 2 == 0 }"
      }
    }
  }
}
```

I noticed that the issue seemed to be ClassNotFound errors related to the security manager, so I started adding permissions until I hit a wall because it could not access the generated closure's class.

```
  permission org.elasticsearch.script.ClassPermission "groovy.lang.Closure";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.reflection.CachedMethod";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.reflection.ClassInfo";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.runtime.DefaultGroovyMethodsSupport";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.runtime.GeneratedClosure";
```

cc @rasroh

### PR DESCRIPTION
**Pull request by jasontedor**
The purpose behind this pull request is to add some permissions that Groovy needs to use closures in its current state. While granting `java.lang.reflect.ReflectPermission "supressAccessChecks"` is undesirable, there is an upstream pull request in apache/groovy#248 to modify the Groovy compiler so that these permissions are not necessary. Therefore, the long-term plan is to grant these permissions so as to not break existing scripts, and remove these permissions when the upstream pull request is accepted (and make a breaking change in a major release if not). These permissions are not staying around forever.

Closes #16194

### CODE REVIEW

File: modules/lang-groovy/src/main/plugin-metadata/plugin-security.policy [hunk 1]
```diff
@@ -25,6 +25,7 @@ grant {
   // needed by groovy engine
   permission java.lang.RuntimePermission "accessDeclaredMembers";
   permission java.lang.RuntimePermission "accessClassInPackage.sun.reflect";
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```
Comment (rjernst) [inline modules/lang-groovy/src/main/plugin-metadata/plugin-security.policy]: @rmuir should look at this...there was an enormous amount of work that went into removing this permission across elasticsearch.

**Comment by rmuir** [comment]
There is also a huge historical security risk. I don't think we should do this at all: closures aren't necessary for the kind of scripting happening here. This scripting is *deadly* and I don't think we should add even one single permission beyond what is absolutely required.

**Comment by rmuir** [comment]
Just to put another way, to justify my -1 technically:

Closures have not been supported for **months** in master or 2.2: nobody has noticed. There are **hundreds** of tests that use and abuse groovy, not a single one uses a closure. This speaks volumes: they simply aren't necessary.

The `suppressAccessChecks` is over the top here, it has wide implications. Taking on this enormous risk just to add some unnecessary syntactic sugar is the wrong tradeoff.

**Comment by nik9000** [comment]
If we can't get closures without a suppressacesschecks can we throw a
meaningful error message?

Can we make a test with a complex groovy script that we can point people
that want closures to to say "sorry, we can't support closures because
groovy needs dangerous permissions for them to work but if you structure
your script like this it'll work."

Is that good enough given painless's progress?

**Comment by rmuir** [comment]
We discussed this after @ywelsch dug into exactly why groovy is in this situation. He has a potential patch to improve the situation in groovy.

Today we have no way to separate "basic scripting users" from "advanced users" (people who choose to use groovy for a reason, e.g. because they like features of it). The two different use cases will cause tension for a while.

But for now, we can optimistically commit this, with the intent we will "remove" it again one way or another: best case, our fix to groovy gets accepted, we upgrade, and then its gone. worst case, it takes longer, until groovy becomes "opt-in" as a plugin or whatever (and we have a different default language).

I do worry about subsequent commits that will enable more "long-tail" features of groovy, e.g. `@Grab` or similar, in the meantime we need to be careful and judicious to avoid relying on this permission as its really no good at all. And we should expect plenty of "long-tail" feature issues just like this to be opened after 2.2.

**Comment by jasontedor** [comment]
Relates apache/groovy#248.

**Comment by ywelsch** [comment]
@jasontedor LGTM

**Comment by jasontedor** [comment]
Integrated via bdddea2dd0ecd3624e5484b3a34e668ca74cd6d0.

**Pull request closed without merging**